File: java/leetcode_grind/Day448.java
package leetcode_grind;

import java.util.Arrays;

public class Day448 {
    // https://leetcode.com/problems/partition-array-for-maximum-sum/description
    static class Solution1 {
        public int maxSumAfterPartitioning(int[] arr, int k) {
            int N = arr.length;
            int[] dp = new int[N];
            Arrays.fill(dp, -1);

            var rec = new Object() {
                int apply(int start) {
                    if (start >= N) {
                        return 0;
                    }
                    if (dp[start] > -1) {
                        return dp[start];
                    }

                    int currMax = 0, ans = 0;
                    int end = Math.min(N, start + k);
                    for (int i = start; i < end; i++) {
                        currMax = Math.max(currMax, arr[i]);
                        ans = Math.max(ans, currMax * (i - start + 1) + apply(i + 1));
                    }
                    return dp[start] = ans;
                }
            };

            return rec.apply(0);
        }
    }

}

File: src/leetcode_grind/day_448.rs
    pub fn max_sum_after_partitioning(arr: Vec<i32>, k: i32) -> i32 {
        let n = arr.len();
        let mut dp = vec![0; n + 1];

        for start in (0..n).rev() {
            let mut mx = 0;
            let end = (start + k as usize).min(n);

            for i in start..end {
                mx = mx.max(arr[i]);
                dp[start] = dp[start].max(dp[i+1] + mx * (i - start + 1) as i32);
            }
        }
        dp[0]
    }